File: src/market_micstr_lab/book/order_book.py
from decimal import Decimal
from zlib import crc32


def _kraken_checksum_value(value: Decimal) -> str:
    normalized = format(value, "f").replace(".", "").lstrip("0")
    return normalized or "0"


class OrderBook:

    def __init__(self) -> None:
        self.bids: dict[Decimal, Decimal] = {}
        self.asks: dict[Decimal, Decimal] = {}


    def apply(self, event: dict) -> None:
        if event["channel"] != "book":
            raise ValueError(f"Expected book event, got {event['channel']}")

        if event["event_type"] == "snapshot":
            self.bids.clear()
            self.asks.clear()

        if event["event_type"] not in {"snapshot", "update"}:
            raise ValueError(f"Unsupported event type: {event['event_type']}")

        self._apply_levels(self.bids, event.get("bids", []))
        self._apply_levels(self.asks, event.get("asks", []))
        
        

    def _apply_levels(self, side: dict[Decimal, Decimal], levels: list[list[str]]) -> None:
        for price_str, qty_str in levels:
            price = Decimal(price_str)
            qty = Decimal(qty_str)

            if price <= 0:
                raise ValueError(f"Price must be positive: {price}")

            if qty < 0:
                raise ValueError(f"Quantity cannot be negative: {qty}")

            if qty == 0:
                side.pop(price, None)
            else:
                side[price] = qty



    def best_bid(self) -> tuple[Decimal, Decimal] | None:
        if not self.bids:
            return None

        price = max(self.bids)
        return price, self.bids[price]



    def best_ask(self) -> tuple[Decimal, Decimal] | None:
        if not self.asks:
            return None

        price = min(self.asks)
        return price, self.asks[price]



    def spread(self) -> Decimal | None:
        best_bid = self.best_bid()
        best_ask = self.best_ask()

        if best_bid is None or best_ask is None:
            return None

        return best_ask[0] - best_bid[0]
        
        

    def mid_price(self) -> Decimal | None:
        best_bid = self.best_bid()
        best_ask = self.best_ask()

        if best_bid is None or best_ask is None:
            return None

        return (best_bid[0] + best_ask[0]) / Decimal("2")
        
        
        
    def top_bids(self, n: int) -> list[tuple[Decimal, Decimal]]:
        return [(price, self.bids[price]) for price in sorted(self.bids, reverse=True)[:n]]




    def top_asks(self, n: int) -> list[tuple[Decimal, Decimal]]:
        return [(price, self.asks[price]) for price in sorted(self.asks)[:n]]
        
        
        
    def kraken_checksum(self, depth: int = 10) -> int:
        checksum_parts = []

        for price, qty in self.top_asks(depth):
            checksum_parts.append(_kraken_checksum_value(price))
            checksum_parts.append(_kraken_checksum_value(qty))

        for price, qty in self.top_bids(depth):
            checksum_parts.append(_kraken_checksum_value(price))
            checksum_parts.append(_kraken_checksum_value(qty))

        checksum_input = "".join(checksum_parts).encode("ascii")
        return crc32(checksum_input) & 0xFFFFFFFF


    def checksum_errors(self, event: dict, depth: int = 10) -> list[str]:
        exchange_checksum = event.get("checksum")
        if exchange_checksum is None:
            return []

        try:
            expected_checksum = int(exchange_checksum)
        except (TypeError, ValueError):
            return [f"invalid checksum value: {exchange_checksum}"]

        local_checksum = self.kraken_checksum(depth=depth)
        if local_checksum != expected_checksum:
            return [
                f"checksum mismatch: exchange={expected_checksum}, local={local_checksum}"
            ]

        return []


        
        
    
    def validation_errors(self) -> list[str]:
        errors = []

        if not self.bids:
            errors.append("missing bids")

        if not self.asks:
            errors.append("missing asks")

        best_bid = self.best_bid()
        best_ask = self.best_ask()

        if best_bid is not None and best_ask is not None and best_bid[0] >= best_ask[0]:
            errors.append("crossed book")

        return errors




    def is_valid(self) -> bool:
        return not self.validation_errors()


    def assert_valid(self) -> None:
        errors = self.validation_errors()
        if errors:
            raise ValueError("; ".join(errors))

File: cpp/src/order_book.cpp
#include "mml/order_book.hpp"

namespace mml {

void OrderBook::apply_snapshot(const std::vector<Level>& bids, const std::vector<Level>& asks) {
    bids_.clear();
    asks_.clear();

    for (const auto& level : bids) {
        apply_level(bids_, level);
    }

    for (const auto& level : asks) {
        apply_level(asks_, level);
    }
}

void OrderBook::apply_update(Side side, Level level) {
    if (side == Side::Bid) {
        apply_level(bids_, level);
    } else {
        apply_level(asks_, level);
    }
}

void OrderBook::apply_updates(const std::vector<Level>& bids, const std::vector<Level>& asks) {
    for (const auto& level : bids) {
        apply_update(Side::Bid, level);
    }

    for (const auto& level : asks) {
        apply_update(Side::Ask, level);
    }
}

std::optional<Level> OrderBook::best_bid() const {
    if (bids_.empty()) {
        return std::nullopt;
    }

    const auto& [price, quantity] = *bids_.begin();
    return Level{price, quantity};
}

std::optional<Level> OrderBook::best_ask() const {
    if (asks_.empty()) {
        return std::nullopt;
    }

    const auto& [price, quantity] = *asks_.begin();
    return Level{price, quantity};
}

std::vector<Level> OrderBook::top_bids(std::size_t depth) const {
    std::vector<Level> levels;
    levels.reserve(depth);

    for (const auto& [price, quantity] : bids_) {
        if (levels.size() == depth) {
            break;
        }

        levels.push_back(Level{price, quantity});
    }

    return levels;
}

std::vector<Level> OrderBook::top_asks(std::size_t depth) const {
    std::vector<Level> levels;
    levels.reserve(depth);

    for (const auto& [price, quantity] : asks_) {
        if (levels.size() == depth) {
            break;
        }

        levels.push_back(Level{price, quantity});
    }

    return levels;
}

std::optional<std::int64_t> OrderBook::spread() const {
    const auto bid = best_bid();
    const auto ask = best_ask();

    if (!bid.has_value() || !ask.has_value()) {
        return std::nullopt;
    }

    return ask->price - bid->price;
}

std::optional<double> OrderBook::mid_price() const {
    const auto bid = best_bid();
    const auto ask = best_ask();

    if (!bid.has_value() || !ask.has_value()) {
        return std::nullopt;
    }

    return (static_cast<double>(bid->price) + static_cast<double>(ask->price)) / 2.0;
}

std::vector<std::string> OrderBook::validation_errors() const {
    std::vector<std::string> errors;

    if (bids_.empty()) {
        errors.push_back("missing bids");
    }

    if (asks_.empty()) {
        errors.push_back("missing asks");
    }

    const auto bid = best_bid();
    const auto ask = best_ask();

    if (bid.has_value() && ask.has_value() && bid->price >= ask->price) {
        errors.push_back("crossed book");
    }

    return errors;
}

bool OrderBook::is_valid() const {
    return validation_errors().empty();
}

std::size_t OrderBook::bid_level_count() const {
    return bids_.size();
}

std::size_t OrderBook::ask_level_count() const {
    return asks_.size();
}

}  // namespace mml
